Add domain layer to feedback_services.py

maitbayev opened the pull request:
Fixes #1369

maitbayev commented:
This is an attempt to refactor feedback_services.py to have a domain layer.
In order to finish the issue #1369, following things should be done:
[x] Domain for FeedbackThreadModel and tests 
[ ] Domain for FeedbackThreadMessageModel and tests 
[ ] Domain for FeedbackAnalyticalModel

Can you confirm the first part? 

Also, I have some questions/confustions, related to this and not:
1. How to debug when running backend tests, i.e when you type in command line "bash scripts/run_backend_tests.sh". 
I tried to use the "print" statement, but it didn't work. 
2. What is the "summary" field in FeedbackThreadModel? I couldn't find any place where it is used. 
3. It is confusing, id of FeedbackThreadModel. In some places, thread_id means suffix of the thread.id (i.e "[EXPLORATION_ID].[THREAD_ID]"), and in other places thread_id means thread.id. Thought it is worth sharing.

seanlip commented:
Thanks, @maitbayev! Let me get back to you on this once #1307 is in (it should be quite close now). That PR (and the 'suggestions' branch in general) has some major changes to the feedback functionality, so I think it's worth waiting for it to go through before introducing this. I'll check back on this and comment then.

seanlip commented:
Oh, btw, here are answers to the questions!

(1) Generally, I raise an Exception.
(2) Hm, I'm not sure. It's possible that it's a field that never ended up getting used, as you say (though I'll need to do a production test to see if it's safe to just remove, so let's leave it as it is for now).
(3) @AllanYangZhou and @sbhowmik89 have been working on fixing this in their work on the suggestions feature, I think.

seanlip commented:
Hi @maitbayev, #1307 has gone in now, so I'm taking a look at this again. I think it's very much along the right lines, but there may be some modifications due to #1307. Can you resolve the conflicts, and then I'll take another look and make inline comments?

Thanks!

seanlip commented:
Hi @maitbayev -- I was just going through some old issues, and found this one. What do you think about cleaning up this PR?

maitbayev commented:
Hi @seanlip,
Sorry about not cleaning up this PR a bit earlier and for the late reply, I was super busy. 
Definitely! Next thing, what I will do in my free time. I will try over the weekend.
Thanks for reminding me.

maitbayev commented on core/domain/feedback_services.py:
```diff
@@ -179,7 +182,7 @@ def get_last_updated_time(exploration_id):
     """
-    threadlist = get_all_threads(exploration_id, False)
+    threadlist = get_displayable_threads(exploration_id, False)
     return max(
```
I am not sure what is the has_suggestion,
but shouldn't be here max from all threads regarding of has_suggestion ? 
It means that: max from get_displayable_threads(exploration_id, False) and get_displayable_threads(exploration_id, True)?

seanlip replied:
Yes, that's right. There are two types of feedback threads, those with a suggestion attached, and those without. has_suggestion is a boolean indicating which of these is the case. The maximum here should be taken over all the threads.

seanlip replied:
Ah, I see -- you're saying that the original code was wrong. I think that is the case, actually.

Having said that, I traced this back to the frontend, and I found that this function is only being used by core.controllers.feedback, in a handler that never actually gets called by the frontend. Can you confirm this? If so, maybe we should just get rid of it (and the corresponding unused stuff in the controller, feconf.py and main.py)?

maitbayev replied:
By grepping "feedback_last_updated" couldn't find any front-end code using it as a url. 
So removing it.

seanlip replied:
SGTM. Thanks!

maitbayev commented:
Hi @seanlip,
This commit contains domain for the FeedbackThreadModel and tests,
Let me know whether I am in the right direction.
Also, I have a question/concern which you can find above.
Thanks.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -234,46 +237,63 @@ def get_suggestion(exploration_id, thread_id):
 
-def get_all_threads(exploration_id, has_suggestion):
+def get_displayable_threads(exploration_id, has_suggestion):
     """Return a list of all threads with suggestions."""
```
Hm, looks like the (old) docstring here is wrong -- it should be "Returns a list of all threads with suggestions if has_suggestion is True; otherwise, returns a list of all threads with no suggestions."

seanlip commented on core/domain/feedback_domain.py:
```diff
@@ -0,0 +1,16 @@
+class FeedbackThread(object):
+    """Domain object for an FeedbackThreadModel."""
```
Domain object for a feedback thread.

seanlip commented on core/domain/feedback_domain.py:
```diff
@@ -0,0 +1,16 @@
+class FeedbackThread(object):
```
Add copyright headers, etc.

seanlip commented on core/domain/feedback_jobs_continuous_test.py:
```diff
@@ -400,7 +400,8 @@ def test_thread_closed_job_running(self):
 
             # Trigger close event.
-            threadlist = feedback_services.get_all_threads(exp_id, False)
+            threadlist = feedback_services.get_displayable_threads(exp_id,
```
Suggest

    threadlist = feedback_services.get_displayable_threads(
        exp_id, False)

(and similarly below).

seanlip commented:
This looks very nice, thanks @maitbayev! I have just one question: why change get_all_threads to get_displayable_threads? (I don't have an objection, but I couldn't understand why the change was made in the first place either.)

maitbayev commented:
@seanlip, I renamed get_all_threads to get_displayable_threads, because I've added get_threads method which returns threads of domain instances.  Personally, it was confusing get_all_threads vs get_threads. Thus, I've renamed it to get_displayable_threads, get_closed_threads -> get_displayable_closed_threads and get_open_threads -> get_displayable_open_threads. Later just for naming consistency. 

Alternative name is get_threads_dict, but for me dict meant more direct dict representation, i.e here, there is a field transformation such as original_author_username instead of original_author_id. That's why I've used "displayable" substring. (actually now I don't see any confusion of using git_threads_dict).

Shortly, it was just a personal preference. What do you think?

seanlip commented:
Can we keep get_all_threads() (since 'displayable' implies there are non-displayable threads), and also return thread domain objects? In addition, can we add a to_dict() method to each domain object, which the frontend can call as needed? Then the domain/services layer can deal entirely with domain objects, which seems cleaner.

maitbayev commented on core/domain/feedback_services_test.py:
```diff
@@ -232,7 +232,52 @@ def test_get_all_threads_without_suggestion(self):
                            self._get_milliseconds):
                 threads = feedback_services.get_all_threads(
-                    self.EXP_ID2, False)
+                    self.EXP_ID1, False)
```
I am not against using self.swap, but for curiosity: 
Why we used self.swap to get_threads?
Will it be better to create threads actually and test it?
I think, this tests were wrong logically.

seanlip replied:
Whether the existing tests are correct depends on what the author's intention was. Presumably, when the self.swap() is used, the author intends to test the behaviour of what lies "in between" FeedbackThreadModel.get_threads() and feedback_services.get_all_threads(); when it is not used, the author intends to test the behaviour of feedback_services.get_all_threads().

Having said that, using self.swap() does introduce some skew. I think it would be nicer to remove it, so please feel free to go ahead and do that if it isn't too hard.

maitbayev replied:
I didn't mean that using the self.swap is wrong, it does its job, so it should be fine.
What I meant by logically wrong test is:

fake get_threads(i.e self._get_threads) returns: 
`
['exp1.thread1', 'exp1.thread2', 'exp1.thread3', 'exp2.thread4', 'exp2.thread4']. 
`
After that we try to call:
`
feedback_services.get_all_threads('exp2')
`
Which in a consequence calls get_threads('exp2') which returns fake get_threads.
In a result, we will receive all threads including 'exp1.thread1' and other. 
I think this is logically wrong.
That is why I've changed _get_threads to return only 'exp1.*' and tests to use get_all_threads(exp1.*).

I was able to see the authors intends, but in my opinion current version more correct. 

In the end, I would prefer `for thread in self.thread thread.put()` though.
In later case authors intends would be correct.

seanlip replied:
Oh, I see what you mean. Yep, that is bad, and I agree with what you've done.

In fact, could we go ahead with removing the swap altogether? There's a tradeoff here between testing too much stuff and dealing with hidden dependencies behind the scenes, and I'm more concerned about the latter than the former.

maitbayev commented:
Hi @seanlip,
Yeap, I like it. I've changed accordingly.

maitbayev commented:
Hi @seanlip,
This is full version of the PR, also I have a concern above.
Thanks

seanlip commented on core/controllers/feedback.py:
```diff
@@ -59,11 +59,12 @@ class ThreadHandler(base.BaseHandler):
-                exploration_id, thread_id)})
+            'messages': [m.to_dict() for m in feedback_services.get_messages(
+                exploration_id, thread_id)]})
+        suggestion = feedback_services.get_suggestion(exploration_id, thread_id)
```
Small refactoring suggestion: maybe declare suggestion at the top of get(), and then do a single self.values.update() with both 'messages' and 'suggestion'? Seems cleaner.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -1540,8 +1540,8 @@ def accept_suggestion(editor_id, thread_id, exploration_id, commit_message):
             exploration_id, thread_id)
-        suggestion_author_username = suggestion['author_name']
-        change_list = _create_change_list_from_suggestion(suggestion)
+        suggestion_author_username = suggestion.to_dict()['author_name']
```
This seems odd. I would expect suggestion-dot-something, or suggestion.get_author_name(). The latter probably makes most sense.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -1540,8 +1540,8 @@ def accept_suggestion(editor_id, thread_id, exploration_id, commit_message):
-        suggestion_author_username = suggestion['author_name']
-        change_list = _create_change_list_from_suggestion(suggestion)
+        suggestion_author_username = suggestion.to_dict()['author_name']
+        change_list = _create_change_list_from_suggestion(suggestion.to_dict())
```
Same thing -- there shouldn't be a to_dict() conversion here. Instead, _create_change_list_from_suggestion() should work with Suggestion domain objects.

seanlip commented on core/domain/feedback_domain.py:
```diff
@@ -0,0 +1,129 @@
+# limitations under the License.
+
+"""Domain objects for feedback models"""
+from core.domain import user_services
```
Add single line above this one.

seanlip commented on core/domain/feedback_domain.py:
```diff
@@ -0,0 +1,129 @@
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
```
2016, technically

seanlip commented on core/domain/feedback_domain.py:
```diff
@@ -0,0 +1,129 @@
+
+
+class FeedbackAnalytics(object):
+    """Domain object for a feedback analytics"""
```
Domain object representing feedback analytics for a specific exploration.

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+
+    def test_get_exp_id_from_full_thread_id(self):
+        feedback_thread_cl = feedback_domain.FeedbackThread
+        self.assertEqual(feedback_thread_cl.get_exp_id_from_full_thread_id(
```
Try, instead:

```
        self.assertEqual(
            feedback_domain.FeedbackThread.get_exp_id_from_full_thread_id(
                self.FULL_THREAD_ID),
            self.EXP_ID)
```

or

```
        observed_exp_id = (
            feedback_domain.FeedbackThread.get_exp_id_from_full_thread_id(
                self.FULL_THREAD_ID))
        self.assertEqual(observed_exp_id, self.EXP_ID)
```

seanlip replied:
(It seems a bit odd to just rename the class.)

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+            self.FULL_THREAD_ID), self.EXP_ID)
+
+    def test_get_thread_id_from_full_thread_id(self):
+        feedback_thread_cl = feedback_domain.FeedbackThread
```
Ditto.

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+    def setUp(self):
+        super(FeedbackAnalyticsDomainUnitTests, self).setUp()
+
+    def to_dict(self):
```
This won't fire. Tests should start with test_*.

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+
+    def test_to_dict(self):
+        expected_suggestion = {
```
expected_suggestion_dict

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+            'description': 'a description',
+            'state_content': 'a state content'
+        }
+        suggestion = feedback_domain.Suggestion(
```
observed_suggestion_dict (or suggestion_dict)

maitbayev replied:
renamed it to observed_suggestion, it is a domain object not a dict.

seanlip replied:
oh, sorry! yep, you're right.

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+
+    def test_to_dict(self):
+        fake_date = datetime.datetime(2016, 4, 10, 0, 0, 0, 0)
+        expected_message = {
```
expected_message_dict (without 'dict' this is assumed to be a domain object)

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+            'updated_status': 'open',
+            'updated_subject': 'an updated subject'
+        }
+        message = feedback_domain.FeedbackMessage(
```
message_dict, or observed_message_dict

maitbayev replied:
renamed to observed_message

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+
+    def test_to_dict(self):
+        fake_date = datetime.datetime(2016, 4, 10, 0, 0, 0, 0)
+        expected_thread = {
```
expected_thread_dict

seanlip commented on core/domain/feedback_domain_test.py:
```diff
@@ -0,0 +1,139 @@
+            'subject': u'a subject',
+            'last_updated': utils.get_time_in_millisecs(fake_date)
+        }
+        thread = feedback_domain.FeedbackThread(
```
thread_dict, or observed_thread_dict

maitbayev replied:
renamed to observed_thread

seanlip replied:
Ack, my bad.

seanlip commented on core/domain/feedback_jobs_continuous_test.py:
```diff
@@ -67,7 +67,8 @@ def _run_job_and_check_results(self, exp_id, expected_thread_analytics):
-            ModifiedFeedbackAnalyticsAggregator.get_thread_analytics(exp_id),
+            ModifiedFeedbackAnalyticsAggregator.get_thread_analytics(
+                exp_id).to_dict(),
             expected_thread_analytics)
```
expected_thread_analytics_dict

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -163,13 +163,16 @@ def test_create_and_get_suggestion(self):
         thread = feedback_models.FeedbackThreadModel.get(
             feedback_models.FeedbackThreadModel.generate_full_thread_id(
                 self.EXP_ID3, self.THREAD_ID1))
+        expected_suggestion = {
```
expected_suggestion_dict

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -134,7 +134,7 @@ def setUp(self):
         thread4 = feedback_models.FeedbackThreadModel(
             id=feedback_models.FeedbackThreadModel.generate_full_thread_id(
-                self.EXP_ID2, self.THREAD_ID4),
+                self.EXP_ID1, self.THREAD_ID4),
```
I think the original intention here was to use self.EXP_ID2 -- it's the exploration_id below which is wrong. It may not matter though -- but if we're not using EXP_ID2 at all we should probably remove it and rename EXP_ID3 to EXP_ID2.

maitbayev replied:
So, I've answered below to this note. I will include relevant parts here:

fake get_threads(i.e self._get_threads) returns: 
`
['exp1.thread1', 'exp1.thread2', 'exp1.thread3', 'exp2.thread4', 'exp2.thread4']. 
`
After that we try to call:
`
feedback_services.get_all_threads('exp2')
`
Which, in a consequence calls get_threads('exp2') which returns fake get_threads.
In a result, we will receive all threads including "exp1.\*", but we shouldn't.
Thus, I think this is logically wrong.
That is why I've changed _get_threads to return only "exp1.\*" and tests to use get_all_threads('exp1').

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -232,7 +235,52 @@ def test_get_all_threads_without_suggestion(self):
+        self.assertDictContainsSubset(expected_thread, threads[0].to_dict())
+
+        # Viewer creates feedback thread
+        expected_thread = {
```
expected_thread_dict (we should be careful in distinguishing thread domain objects and their dict representations)

seanlip commented on core/domain/subscription_services_test.py:
```diff
@@ -146,10 +147,11 @@ def test_posting_to_feedback_thread_results_in_subscription(self):
         self.assertEqual(len(thread_ids_subscribed_to), 1)
         full_thread_id = thread_ids_subscribed_to[0]
-        thread_id = feedback_services.get_thread_id_from_full_thread_id(
+        feedback_thread_class = feedback_domain.FeedbackThread
```
See above -- I think it's a bit nicer to avoid just aliasing a class, if possible. One less name to keep in context.

seanlip commented on core/domain/user_jobs_continuous.py:
```diff
@@ -220,9 +218,10 @@ def map(item):
 
         for feedback_thread_id in feedback_thread_ids_list:
-            exp_id = feedback_services.get_exp_id_from_full_thread_id(
+            feedback_thread_cl = feedback_domain.FeedbackThread
```
ditto

seanlip commented:
Thanks, @maitbayev! I took a first pass, and it looks good.

One question: could you please test the feedback/suggestion system manually (fairly carefully) to ensure there's no breakage? Unfortunately we have poor e2e test coverage for this. (We have an issue open -- #1450 -- for e2e tests for feedback/suggestions; do you think you might be interested in taking this on at some point?)

maitbayev commented:
Hi @seanlip,
I addressed all of your comments. 
I will update when I will check it manually.
Sure, I can try writing end2end tests for suggestion.

maitbayev commented:
Hi @seanlip,
I am confirming that I checked it manually. Checks done:
1. Createe feedback Thread, anonymous and not
2. Created suggestions 
3. Suggestions accepted, rejected
4. Feedback message changed status, tried different statuses

seanlip commented:
Hi @maitbayev -- this looks really nice! The only remaining point is the test you identified as being problematic -- I think we should fix it and that it's better to remove the swap if possible. Would you mind doing that? Once you're happy with it, this PR can be merged.

Also, I'm not sure what's going on with codecov -- it's become more noisy lately. Perhaps they pushed an update or something. I'll try and investigate, but you can ignore it in the meantime.

Thanks!

maitbayev commented:
Hi @seanlip,
I removed swaps from feedback_services_test.

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -87,11 +86,13 @@ class SuggestionQueriesUnitTests(test_utils.GenericTestBase):
 
-    def _get_thread_with_suggestions(self, unused_exploration_id):
-        return self.threads_with_suggestions
+    def _put_thread_with_suggestions(self):
```
nit: s/thread/threads

seanlip replied:
Edit: see other comment.

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -87,11 +86,13 @@ class SuggestionQueriesUnitTests(test_utils.GenericTestBase):
 
-    def _get_threads(self, unused_exploration_id):
-        return self.threads
+    def _put_threads(self):
```
_put_all_threads(), perhaps

seanlip replied:
Actually -- instead of all this, can we just do "for thread in self.threads:\n thread.put()" on setUp()? Then we don't need to repeat it all the time.

seanlip replied:
And maybe move test_create_and_get_suggestion() to a new suite or bundle it with one of the other suites -- or alternatively modify it so that it doesn't interfere with what's already been done in setUp().

maitbayev replied:
The reason was one test using "_put_threads_with_suggestions", but I don't see any difference using _put_threads instead.

seanlip replied:
Yep, agreed.

maitbayev replied:
Can you explain a bit more about your last comment ?

seanlip replied:
Ah -- I was wondering why you put _put_threads() everywhere instead of putting it in setUp(). I guessed that maybe it was because it would cause problems with test_create_and_get_suggestion() -- hence my last comment. However, if it doesn't, it's fine to leave it as-is!

seanlip commented:
Thanks -- reviewed!

maitbayev commented:
G

seanlip commented:
LGTM. Thanks @maitbayev! We can merge this once the Travis checks pass.

maitbayev commented:
@seanlip,
I accidentally closed this PR and typed one letter message in my phone, which seems like isn't revertable in github phone web interface.
Anyway, from your last comment, the last commit should address your comments.

seanlip commented:
Yep, no problem -- I've reviewed it, and it looks great. Thanks, @maitbayev!
